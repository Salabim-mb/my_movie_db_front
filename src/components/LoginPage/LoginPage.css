.card__container {
    margin: 10rem 25%;
    vertical-align: middle;
}

.card__header {
    text-align: center;
}

@media(max-width: 768px) {
    .card__container {
        margin: auto 5%;
    }
}