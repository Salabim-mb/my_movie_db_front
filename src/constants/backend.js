const backend_path = process.env.REACT_APP_BACKEND_PATH;

export const backend = {
    MOVIES: backend_path + "movies/",
    USERS: backend_path + "user/",
    PLAIN: backend_path,

};