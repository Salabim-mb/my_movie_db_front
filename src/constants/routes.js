import MovieList from "../components/MovieList";

const paths = {
    MAIN: "/",
    LOGIN: "/login",
    REGISTER: "/register",
    MOVIES: "/movies",
    NEW_MOVIE: "/movies/new"
};

export default [
    {
        path: paths.MOVIES,
        exact: true,
        component: MovieList
    }r
]

export {paths};