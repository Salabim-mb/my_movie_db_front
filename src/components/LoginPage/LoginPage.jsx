import React, {useContext, useState} from "react";
import {Alert, Button, Card, Col, Form, Row} from "react-bootstrap";
import {backend} from "../../constants/backend";
import {getHeaders} from "../../utils/CORSHeaders";
import "./LoginPage.css";
import {IndexLinkContainer} from 'react-router-bootstrap';
import {paths} from "../../constants/routes";
import {Redirect} from "react-router";
import {UserContext} from "../../context/UserContext";

const loginUser = async (data) => {
    let url = `${backend.LOGIN}`;
    let headers = getHeaders();

    let res = await fetch(url, {
        method: "POST",
        headers,
        body: JSON.stringify({
            userEmail: data.email,
            userPassword: data.password
        })
    });

    if (res.status === 200 || res.status === 201) {
        return await res.json();
    } else {
        throw res.status;
    }
};

const LoginPage = () => {
    const [data, setData] = useState({
        email: "",
        password: "",
    });
    const [disabled, setDisabled] = useState(false);
    const [error, setError] = useState(false);
    const [validated, setValidated] = useState(false);
    const [correct, setCorrect] = useState(false);
    const [redirect, setRedirect] = useState(false);

    const user = useContext(UserContext);

    const submitForm = async (e) => {
        e.preventDefault();
        setError(false);
        setValidated(true);
        if (e.currentTarget.checkValidity() === false) {
            e.stopPropagation();
        } else {
            setDisabled(true);
            try {
                let {userToken, userName, userSurname, userEmail} = await loginUser(data);
                user.login(userToken, {
                    name: userName,
                    surname: userSurname,
                    email: userEmail
                });
                setCorrect(true);
                setTimeout(() => setRedirect(true), 3000);
            } catch(e) {
                console.log(e);
                setError(true);
            } finally {
                setDisabled(false);
            }
        }
    }

    return (
        <Card className="card__container">
            <Card.Header className="card__header">
                <Card.Title as="h3">Sign in</Card.Title>
            </Card.Header>
            <Card.Body>
                <Form noValidate validated={validated} onSubmit={submitForm}>
                    <Form.Group controlId="email">
                        <Form.Label>Email</Form.Label>
                        <Form.Control
                            type="email"
                            required
                            placeholder="johndoe@example.com"
                            value={data.email}
                            onChange={e => setData({...data, email: e.target.value})}
                        />
                        <Form.Control.Feedback type="invalid">
                            Correct email must be provided
                        </Form.Control.Feedback>
                    </Form.Group>
                    <Form.Group controlId="password">
                        <Form.Label>Password</Form.Label>
                        <Form.Control
                            type="password"
                            minLength={8}
                            required
                            placeholder="Your password"
                            name="password"
                            value={data.password}
                            onChange={e => setData({...data, password: e.target.value})}
                        />
                        <Form.Control.Feedback type="invalid">
                            Correct password must be provided
                        </Form.Control.Feedback>
                    </Form.Group>
                    <div className="justify-content-end d-flex">
                        <Button
                            type="submit"
                            disabled={disabled}
                            variant="success"
                        >
                            {disabled ? "Loading..." : "Log in!"}
                        </Button>
                    </div>
                    <IndexLinkContainer to={paths.REGISTER}>
                        <Card.Link>Don't have an account? Sign up!</Card.Link>
                    </IndexLinkContainer>
                    {error && <Alert variant="danger">Something went wrong while trying to login user</Alert>}
                    {correct && <Alert variant="success">Login successful. Redirecting...</Alert>}
                    {redirect && <Redirect to={paths.MAIN}/>}
                </Form>
            </Card.Body>
        </Card>
    )
};

export default LoginPage;