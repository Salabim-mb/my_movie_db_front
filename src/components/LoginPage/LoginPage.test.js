import React from 'react';
import {render, fireEvent, waitForElement} from '@testing-library/react';
import {createMemoryHistory} from 'history';
import LoginPage from "./LoginPage";
import {MemoryRouter} from "react-router";
import {Router} from "react-router-dom";
import {paths} from "../../constants/routes";

const renderWithRouter = (
    ui,
    {
        route = "/login",
        history = createMemoryHistory({ initialEntries: [route] })
    } = {}
) => {
    return {
        ...render(
            <Router history={history}>{ui}</Router>
        ),
        history
    }
};

describe("LoginPage", () => {
    let apiFail;
    let token;
    let data;
    let invalid;

    beforeAll(() => {
        global.fetch = jest.fn().mockImplementation((input, init) => {
            return new Promise((resolve, reject) => {
                if (apiFail) {
                    resolve({status: 500})
                } else {
                    if (invalid) {
                        resolve({status: 400});
                    } else {
                        resolve({status: 200, json: () => Promise.resolve(token)})
                    }
                }
            });
        });
    });

    beforeEach(() => {
        invalid = false;
        apiFail = false;
        jest.clearAllMocks();
        token = "123";
        data = {
            email: "qwe@qwe.qwe",
            password: "zaq1@WSX"
        };
    });

    it("should match snapshot", () => {
        const {container} = render(
            <MemoryRouter>
                <LoginPage/>
            </MemoryRouter>
        );

        expect(container).toMatchSnapshot();
    });

    it("should log in correctly", async () => {
        const {history, getByLabelText, getByText} = renderWithRouter(
            <LoginPage />
        );

        fireEvent.change(getByLabelText("Email"), {
            target: {value: data.email}
        });
        fireEvent.change(getByLabelText("Password"), {
            target: {value: data.password}
        });
        fireEvent.click(getByText("Log in!"));

        await waitForElement(() => getByText("Login successful", {exact: false}));

        expect(getByText("Login successful", {exact: false})).toBeInTheDocument();
        await expect(history.location.pathname).toBe(paths.LOGIN);

    });

    it("should render error on invalid credentials", async () => {
        invalid = true;
        const {history, getByLabelText, getByText} = renderWithRouter(
            <LoginPage />
        );

        fireEvent.change(getByLabelText("Email"), {
            target: {value: data.email}
        });
        fireEvent.change(getByLabelText("Password"), {
            target: {value: data.password}
        });
        fireEvent.click(getByText("Log in!"));

        await waitForElement(() => getByText("Something went wrong", {exact: false}));

        expect(getByText("Something went wrong", {exact: false})).toBeInTheDocument();
        await expect(history.location.pathname).not.toBe(paths.USER_PAGE);
    });

    it("should render alert on api fail", async () => {
        apiFail = true;
        const {history, getByLabelText, getByText} = renderWithRouter(
            <LoginPage />
        );

        fireEvent.change(getByLabelText("Email"), {
            target: {value: data.email}
        });
        fireEvent.change(getByLabelText("Password"), {
            target: {value: data.password}
        });
        fireEvent.click(getByText("Log in!"));

        await waitForElement(() => getByText("Something went wrong", {exact: false}));

        expect(getByText("Something went wrong", {exact: false})).toBeInTheDocument();
        await expect(history.location.pathname).not.toBe(paths.USER_PAGE);
    });

    it("should redirect to register page", () => {
        const {history, getByText} = renderWithRouter(
            <LoginPage />
        );
        fireEvent.click(getByText("Don't have an account", {exact: false}));
        expect(history.location.pathname).toBe(paths.REGISTER)
    });
});