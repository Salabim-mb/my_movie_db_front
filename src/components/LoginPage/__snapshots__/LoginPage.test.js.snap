// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage should match snapshot 1`] = `
<div>
  <div
    class="card__container card"
  >
    <div
      class="card__header card-header"
    >
      <h3
        class="card-title"
      >
        Sign in
      </h3>
    </div>
    <div
      class="card-body"
    >
      <form
        class=""
        novalidate=""
      >
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="email"
          >
            Email
          </label>
          <input
            class="form-control"
            id="email"
            placeholder="johndoe@example.com"
            required=""
            type="email"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Correct email must be provided
          </div>
        </div>
        <div
          class="form-group"
        >
          <label
            class="form-label"
            for="password"
          >
            Password
          </label>
          <input
            class="form-control"
            id="password"
            minlength="8"
            name="password"
            placeholder="Your password"
            required=""
            type="password"
            value=""
          />
          <div
            class="invalid-feedback"
          >
            Correct password must be provided
          </div>
        </div>
        <div
          class="justify-content-end d-flex"
        >
          <button
            class="btn btn-success"
            type="submit"
          >
            Log in!
          </button>
        </div>
        <a
          class="card-link"
          href="/register"
        >
          Don't have an account? Sign up!
        </a>
      </form>
    </div>
  </div>
</div>
`;
